proxy/internal/modules/store/mysql.go

```diff
@@ -0,0 +1,496 @@
+		ExecContext(ctx, add.Mod.Source, add.Mod.Version, []byte(add.ModFile), newRevInfo(add.Mod).Bytes(), add.UniqueID)
+	if err != nil {
+		m.emitter.Count("db-insert-mod-failure", 1)
+	} else {
+		m.emitter.GaugeMS("db-insert-mod-elapsed-ms", start)
+	}
+
+	return err
+}
+
+func (m *mysqlStore) IDs() (Ranges, error) {
+	m.log.Tracef("retrieving all ids")
+
+	start := time.Now()
+	ids, err := m.ids()
+	if err != nil {
+		return nil, err
+	}
+
+	m.emitter.GaugeMS("db-get-ids-elapsed-ms", start)
+	return ranges(ids), nil
+}
+
+func (m *mysqlStore) Summary() (int, int, error) {
+	m.log.Tracef("retrieving all sources and computing summary")
+	return m.countSourcesAndVersions()
+}
+
+const (
+	insertModuleContentsSQL = iota
+	selectModuleContentsSQL
+	moduleContentsExistsSQL
+	deleteModuleContentsSQL
+	insertModuleSQL
+	selectRegistryIDSQL
+	selectAllRegistryIDsSQL
+	countVersionsSQL
+	selectModuleRevInfoSQL
+	selectGoModContentsSQL
+	selectModuleVersionsSQL
+	updateRegistryIDSQL
+	deleteModuleSQL
+)
+
+type statements map[int]*sql.Stmt
+
+func load(kind string, db *sql.DB) (statements, error) {
+	loaded := make(statements, len(mySQLTexts))
+
+	stmtTexts := mySQLTexts
+	if kind == "postgres" {
+		return loaded, errors.New("postgres is not supported (issue #103)")
+		// stmtTexts = postgreSQLTexts
+	}
+
+	for id, text := range stmtTexts {
+		stmt, err := db.Prepare(text)
+		if err != nil {
+			return nil, errors.Wrapf(err, "bad sql statement: %q", text)
+		}
+		loaded[id] = stmt
+	}
+
+	return loaded, nil
+}
+
+var (
+	mySQLTexts = map[int]string{
+		// modules
+		insertModuleContentsSQL: `insert into proxy_module_contents(path, contents) values (?, ?)`,
+		selectModuleContentsSQL: `select contents from proxy_module_contents where path=?`,
+		moduleContentsExistsSQL: `select count(id) from proxy_module_contents where path=?`,
+		deleteModuleContentsSQL: `delete from proxy_module_contents where path=?`,
+		// index
+		insertModuleSQL:         `insert into proxy_modules(source, version, go_mod_contents, rev_info_contents, registry_mod_id) values (?, ?, ?, ?, ?)`,
+		selectRegistryIDSQL:     `select registry_mod_id from proxy_modules where source=? and version=?`,
+		selectAllRegistryIDsSQL: `select registry_mod_id from proxy_modules`,
+		countVersionsSQL:        `select count(version) from proxy_modules group by source`,
+		selectModuleRevInfoSQL:  `select rev_info_contents from proxy_modules where source=? and version=?`,
+		selectGoModContentsSQL:  `select go_mod_contents from proxy_modules where source=? and version=?`,
+		selectModuleVersionsSQL: `select version from proxy_modules where source=?`,
+		updateRegistryIDSQL:     `update proxy_modules set registry_mod_id=? where source=? and version=?`,
+		deleteModuleSQL:         `delete from proxy_modules where source=? and version=?`,
+	}
+)
+
+func (m *mysqlStore) insertModulesContents(mod coordinates.Module, blob repository.Blob) error {
+	exists, err := m.modelContentsExists(mod)
```

Comment:
this check needs to be in a transaction

Reply:
actually we don't need the check at all; if the insert fails because it's already there, it's the same as having done this check anyway